File: solutions/91.py
class Solution:
    def numDecodings(self, s: str) -> int:
        if not s or s[0] == '0':
            return 0
        last = 1
        prelast = (1, 0)
        for c in s:
            new_last = 0
            if 10 <= prelast[1] * 10 + int(c) <= 26:
                new_last += prelast[0]
            if int(c):
                new_last += last
            prelast = (last, int(c))
            last = new_last
        return last

File: solutions/91.cpp
#include <string>
#include <utility>

class Solution {
 public:
  int numDecodings(std::string s) {
    // track how number of ways to decode will change by appending single digit
    std::pair<int, char> prelast{1, '0'};
    int last = 1;
    for (auto c : s) {
      int new_last = 0;
      // append digit to s[0: i - 2] and test if (s[i-1]s[i]) maps to letter
      if (prelast.second == '1' or (prelast.second == '2' and c < '7')) {
        new_last += prelast.first;
      }
      // append digit to s[0: i - 1] and test if (s[i]) maps to letter
      if (c != '0') {
        new_last += last;
      }
      prelast = {last, c};
      last = new_last;
    }
    return last;
  }
};
